ddtrace/llmobs/_integrations/langchain.py

```diff
@@ -109,6 +178,174 @@ def _llmobs_set_tags(
         elif operation == "tool":
             self._llmobs_set_meta_tags_from_tool(span, tool_inputs=kwargs, tool_output=response)
 
+    def _set_links(self, span: Span):
+        """
+        Sets span links for the given LangChain span, by doing the following:
+        1. Determine the default invoker (parent llmobs span) and span linkage attributes (input-to-input)
+        2. If the instance associated with the span is a step in a chain, look for the last traced step in the chain
+            and establish an output-to-input relationship
+        3. Set the span link on the span from the invoker
+        4. Set the span link on the invoker from the span as an output-to-output relationship
+            a. If the span is a step in the chain, we overwrite the existing output-to-output relationship,
+                    until the last traced step is not overwritten
+                i. This assumes only one output-to-output link for an individual chain
+            b. If the span is not a step in the chain, we add the output-to-output relationship on top of
+                the existing links, even if there is another output-to-output link
+        """
+        instance = self._instances.get(span)
+        if not instance:
+            return
+
+        instance = _extract_bound(instance)
+        parent_span = _get_nearest_llmobs_ancestor(span)
+
+        prev_traced_step_idx = self._set_input_links(instance, span, parent_span)
```

Comment:
For clarity - can we extract out the functionality of finding the previous step IDX and setting the input links into separate functions?

Reply:
ah yes actually that's a good idea, lemme do that